File: lambdakal/examples/Primes.hs
module Primes where

divisibleAny :: Int -> Int -> Bool
divisibleAny n i
  | i < 2     = False
  | n `mod` i == 0 = True
  | otherwise = divisibleAny n (i - 1)

isPrime :: Int -> Bool
isPrime n = not (divisibleAny n (n - 1))

primesUntil :: Int -> [Int]
primesUntil n
  | n < 2     = []
  | isPrime n = n : primesUntil (n - 1)
  | otherwise = primesUntil (n - 1)

main :: IO ()
main = do
  let primes = primesUntil 10000
  print primes

File: lambdakal/examples/primes.py
import sys
sys.setrecursionlimit(50000)

def divisible_any(n: int, i: int) -> bool:
    if i < 2:
        return False
    if n % i == 0:
        return True
    return divisible_any(n, i - 1)

def is_prime(n: int) -> bool:
    return not divisible_any(n, n - 1)

def primes_until(n: int) -> list[int]:
    if n < 2:
        return []
    if not divisible_any(n, n - 1):
        return [n] + primes_until(n - 1)
    return primes_until(n - 1)

def main():
    primes = primes_until(25000)
    print(primes)

if __name__ == "__main__":
    main()